[14.0][ADD] web_assets_warmup

jbaudoux opened the pull request:
Ensure that assets are generated and stored in the DB when Odoo starts

If the assets from the database are not up-to-date, they are regenerated by
Odoo when we print a report, but to do so Odoo forces the commit, so if an
exception occurs after (or during) the report rendering, it let the database in
a broken state (picking have been validated in this case).

To prevent this issue, we need to ensure that the assets are well-generated
when Odoo starts, not when the report is printed.

cc @sebalix @mt-software-de

jbaudoux committed:
[ADD] web_assets_warmup

simahawk commented:
FTR I'm not sure if the name is really explicative. Something like `web_assets_warmup` might be more appropriated. Not a blocker tho.

mt-software-de commented on web_generate_assets/hooks.py:
```diff
@@ -0,0 +1,52 @@
+    reg = registry(dbname)
+    with reg.cursor() as cr:
+        cron_module, cron_ref = "web_generate_assets", "cron_generate_assets"
+        query = """
+           SELECT model, res_id
+           FROM ir_model_data
+           WHERE module=%s
+           AND name=%s;
+        """
```
Since we could have dedicated services for odoo web/odoo cronjobs, we should a a lock.
Otherwise we may get a "serialize access" error. 
```suggestion
        query = """
           SELECT model, res_id
           FROM ir_model_data
           WHERE module=%s
           AND name=%s
           FOR UPDATE;
        """
```

mt-software-de replied:
I don't think we should use a "FOR UPDATE" because it locks the complete ir_model_data table. :-)

mt-software-de committed:
[IMP] web_assets_warmup: Add database lock in post_load_hook

With dedicated services for odoo web and odoo cron
we can get a serialize access error on startup

mt-software-de commented on web_generate_assets/hooks.py:
```diff
@@ -0,0 +1,52 @@
+# Copyright 2020 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
+
+import logging
+import os
+
+from odoo import fields, registry
+
+logger = logging.getLogger(__name__)
+
+
+def active_cron_assets():
+    """Plan the next execution of the cron responsible to generate assets."""
+    if os.environ.get("RUNNING_ENV") == "dev":
+        return
+    dbname = os.environ.get("DB_NAME")
+    reg = registry(dbname)
+    with reg.cursor() as cr:
+        cron_module, cron_ref = "web_generate_assets", "cron_generate_assets"
+        query = """
+           SELECT model, res_id
+           FROM ir_model_data
+           WHERE module=%s
+           AND name=%s;
+        """
+        args = (cron_module, cron_ref)
+        cr.execute(query, args)
+        row = cr.fetchone()
+        # post_load hook is called before the update of the module so the
+        # ir_cron record doesn't exist on first install
+        if row:
+            model, res_id = row
+            if model != "ir.cron":
+                return
```
```suggestion
import hashlib
import logging
import os
import struct

from odoo import fields, registry

logger = logging.getLogger(__name__)


def active_cron_assets():
    """Plan the next execution of the cron responsible to generate assets."""
    if os.environ.get("RUNNING_ENV") == "dev":
        return
    dbname = os.environ.get("DB_NAME")
    reg = registry(dbname)
    with reg.cursor() as cr:
        cron_module, cron_ref = "web_generate_assets", "cron_generate_assets"
        query = """
           SELECT model, res_id
           FROM ir_model_data
           WHERE module=%s
           AND name=%s;
        """
        args = (cron_module, cron_ref)
        cr.execute(query, args)
        row = cr.fetchone()
        # post_load hook is called before the update of the module so the
        # ir_cron record doesn't exist on first install
        if row:
            model, res_id = row
            if model != "ir.cron":
                return
            hasher = hashlib.sha1(str("web_generate_assets").encode())
            int_lock = struct.unpack("q", hasher.digest()[:8])
            cr.execute("SELECT pg_advisory_xact_lock(%s);", (int_lock,))
```

mt-software-de replied:
Since we could have dedicated odoo services for web or cronjobs, we should create a lock. Otherwise we could get a serialize access error.

mt-software-de reviewed (changes requested):
Since we could have dedicated services for web and cronjobs. We should add a lock. Otherwise we can get a "serialize access" error. 

https://github.com/camptocamp/web/pull/3

rvalyi commented:
> FTR I'm not sure if the name is really explicative. Something like `web_assets_warmup` might be more appropriated. Not a blocker tho.

+1 for web_assets_warmup

jbaudoux commented:
web_generate_assets_initially ?

rvalyi commented:
> web_generate_assets_initially ?

asset "cache warmup" is really the technical term web developers would expect in a web framework IMHO. See for instance https://www.google.com/search?client=ubuntu-chr&hs=mW0&sca_esv=573172220&q=web+asset+cache+%22warmup%22+rails&oq=web+asset+cache+%22warmup%22+rails&aqs=heirloom-srp..

pedrobaeza commented:
@bofiltd putting a lot of reviews in few minutes in the corresponding PRs is not trustful and can't be taken as a serious review, so please abstain for doing that.

jbaudoux commented:
Module renamed to **web_assets_warmup**

simahawk commented on web_assets_warmup/hooks.py:
```diff
@@ -0,0 +1,70 @@
+    """Plan the next execution of the cron responsible to generate assets."""
+    if os.environ.get("RUNNING_ENV") == "dev":
+        return
+    dbname = os.environ.get("DB_NAME")
```
`DB_NAME` is not a core odoo var. The db name should be take from config.
Not blocking, you can fix this later as this one is waiting since quite some time.

simahawk commented:
/ocabot merge nobump
